File: docs/snippets/all/views/tensor.py
"""Use a blueprint to show a tensor view."""

import numpy as np
import rerun as rr
import rerun.blueprint as rrb

rr.init("rerun_example_tensor", spawn=True)

tensor = np.random.randint(0, 256, (32, 240, 320, 3), dtype=np.uint8)
rr.log("tensor", rr.Tensor(tensor, dim_names=("batch", "x", "y", "channel")))

blueprint = rrb.Blueprint(
    rrb.TensorView(
        origin="tensor",
        name="Tensor",
        # Explicitly pick which dimensions to show.
        slice_selection=rrb.TensorSliceSelection(
            # Use the first dimension as width.
            width=1,
            # Use the second dimension as height and invert it.
            height=rr.TensorDimensionSelection(dimension=2, invert=True),
            # Set which indices to show for the other dimensions.
            indices=[
                rr.TensorDimensionIndexSelection(dimension=2, index=4),
                rr.TensorDimensionIndexSelection(dimension=3, index=5),
            ],
            # Show a slider for dimension 2 only. If not specified, all dimensions in `indices` will have sliders.
            slider=[2],
        ),
        # Set a scalar mapping with a custom colormap, gamma and magnification filter.
        scalar_mapping=rrb.TensorScalarMapping(colormap="turbo", gamma=1.5, mag_filter="linear"),
        # Fill the view, ignoring aspect ratio.
        view_fit="fill",
    ),
    collapse_panels=True,
)
rr.send_blueprint(blueprint)

File: docs/snippets/all/views/tensor.rs
//! Use a blueprint to show a tensor view.

use ndarray::Array;
use rerun::{
    blueprint::{
        Blueprint, TensorView, archetypes as blueprint_archetypes, components,
    },
    components::{Colormap, MagnificationFilter},
    datatypes::TensorDimensionSelection,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let blueprint = Blueprint::new(
        TensorView::new("Tensor")
            .with_origin("tensor")
            .with_slice_selection(
                &blueprint_archetypes::TensorSliceSelection::new()
                    .with_width(1)
                    .with_height(TensorDimensionSelection {
                        dimension: 2,
                        invert: true,
                    })
                    .with_indices([
                        rerun::components::TensorDimensionIndexSelection::new(
                            2, 4,
                        ),
                        rerun::components::TensorDimensionIndexSelection::new(
                            3, 5,
                        ),
                    ])
                    .with_slider([2]),
            )
            .with_scalar_mapping(
                &blueprint_archetypes::TensorScalarMapping::new()
                    .with_colormap(Colormap::Turbo)
                    .with_gamma(1.5)
                    .with_mag_filter(MagnificationFilter::Linear),
            )
            .with_view_fit(
                &blueprint_archetypes::TensorViewFit::new()
                    .with_scaling(components::ViewFit::Fill),
            ),
    );

    let rec = rerun::RecordingStreamBuilder::new("rerun_example_tensor")
        .with_blueprint(blueprint)
        .spawn()?;

    let tensor =
        Array::from_shape_fn((32, 240, 320, 3), |(batch, x, y, channel)| {
            (batch * 17 + x * 5 + y * 3 + channel * 97) as u8
        });
    rec.log(
        "tensor",
        &rerun::Tensor::try_from(tensor)?
            .with_dim_names(["batch", "x", "y", "channel"]),
    )?;

    Ok(())
}
